lib/utils/addSemicolonForEditInfo.mjs

```diff
@@ -0,0 +1,30 @@
+/** @import { Node as PostcssNode } from 'postcss' */
+
+/**
+ * Adjust the text in EditInfo to include a semicolon when needed.
+ *
+ * @param {PostcssNode} node
+ * @param {{range: [number, number], text: string}} fixData
+ * @returns {{range: [number, number], text: string}}
+ */
+export default function addSemicolonForEditInfo(node, fixData) {
+	if (node.type === 'decl') {
+		if (node.parent?.raws.semicolon || node.parent?.last !== node) {
+			return {
+				...fixData,
+				text: `${fixData.text};`,
+			};
+		}
+	}
+
+	if (node.type === 'atrule') {
+		if (!('nodes' in node) && (node.parent?.raws.semicolon || node.parent?.last !== node)) {
+			return {
+				...fixData,
+				text: `${fixData.text};`,
+			};
+		}
+	}
+
+	return fixData;
```

Comment:
I didn't try it out but something like
```js
	switch (node.type) {
		case 'decl':
			break;
		case 'atrule': 
			if (!'nodes' in node) break;
		default:
			return fixData;
	}

	if (node.parent?.raws.semicolon || node.parent?.last !== node) {
		fixData.text = `${fixData.text};`
	}

	return fixData;
```
could be an alternative.

Reply:
I think the `if (!'nodes' in node) break;` is a bit harder to grasp than the current code.

But I also found it difficult to give these conditions a really good shape.

The current code neatly groups the conditions by node type at the cost of some duplication.

Reply:
What about
```js
	switch (node.type) {
		case 'decl': break;
		case 'atrule': {
			const hasBlock = 'nodes' in node;
			if (!hasBlock) break;
		}
		default: return fixData;
	}
	// …
```
?

ref https://drafts.csswg.org/css-syntax/#at-rules

Reply:
That is the same :D

My concern is that using conditional `break` and implicit switch fall through is really difficult to read and understand.

If written in that style I would request it to be changed to something else.

Reply:
What about this
```js
	const isStatementAtRule = node.type === 'atrule' && !'nodes' in node;
	const isDeclaration = node.type === 'decl';

	if (isStatementAtRule || isDeclaration) {
		const isMissingSemicolon = node.parent?.raws.semicolon || node.parent?.last !== node;

		if (isMissingSemicolon) fixData.text = `${fixData.text};`
	}

	return fixData;
```
then?

I think this is better because all conditions are explicit.

Reply:
Side note:

```js
const foo = {false: 1, bar: 2};
!'bar' in foo; // true
```

The parenthesis aren't optional.

------

How about this?

```js
	const isStatementAtRule = node.type === 'atrule' && !('nodes' in node);
	const isDeclaration = node.type === 'decl';

	if (!(isStatementAtRule || isDeclaration)) return fixData;

	const needsSemicolon = node.parent?.raws.semicolon || node.parent?.last !== node;

	if (!needsSemicolon) return fixData;
	
	return {
		...fixData,
		text: `${fixData.text};`,
	};
```

Reply:
If you don't want to mutate `fixData` directly, yes, it will do as well.

Reply:
@Mouvedia @romainmenke Is this discussion still going on? If not, I think we can merge this PR now. Also, we can refactor it later if we find it necessary.

Reply:
Yes, let's refactor later if needed!